### PR TITLE
Add tests for aggregate functions

### PR DESCRIPTION
**Pull request by JamesGuthrie**

### CODE REVIEW

**Comment by JamesGuthrie** [comment]
This change restructures our aggregate function structure and adds tests. I have one more change that I'm working on in the area of schema generation to finish up this refactoring.

File: src/lib.rs
Comment (cevian) [inline src/lib.rs:1]: do we need this file anymore? seems mostly empty now

File: src/aggregates/mod.rs
Comment (cevian) [inline src/aggregates/mod.rs:28]: why are these here and not in prom_delta.rs

File: src/aggregates/mod.rs
Comment (cevian) [inline src/aggregates/mod.rs:1]: should this be called gapfill_delta.rs?

File: src/aggregates/vector_selector.rs
Comment (cevian) [inline src/aggregates/vector_selector.rs:188]: is this function ever hit in the tests? i.e. are we ever using a parallel plan in tests? can we force parallel plans?

File: src/aggregates/vector_selector.rs
Comment (cevian) [inline src/aggregates/vector_selector.rs:77]: is this ever hit in tests?

**Review by cevian — CHANGES_REQUESTED** [review]
Mostly looks good. Some cleanup requested. Also is there an easy/standard way to get code coverage stats on this repo?

File: src/lib.rs
Comment (JamesGuthrie) [reply src/lib.rs:1]: Ya, we do. It serves as the entrypoint into the lib, kinda like a `main.go` does for a go binary.

File: src/aggregates/mod.rs
Comment (JamesGuthrie) [reply src/aggregates/mod.rs:28]: Good question. `prom_delta_final` is used as the `finalfunc` for all aggregates. Perhaps it would make more sense to rename it, but I don't want to break compatibility (yet).

File: src/aggregates/vector_selector.rs
Comment (JamesGuthrie) [reply src/aggregates/vector_selector.rs:188]: It isn't, thanks for catching that! I've added a test which forces execution of a parallel query, which should hit this function.

File: src/aggregates/vector_selector.rs
Comment (JamesGuthrie) [reply src/aggregates/vector_selector.rs:77]: I _think_ (based on my reading of the PG docs on this topic https://www.postgresql.org/docs/14/xaggr.html#XAGGR-PARTIAL-AGGREGATES) that now that we have a parallel execution test, this function should now also be hit.

File: src/aggregates/mod.rs
Comment (JamesGuthrie) [reply src/aggregates/mod.rs:1]: Kinda maybe, but probably not directly. Rust's module system expects there to be a module for `aggregates` somewhere (because we have defined `mod aggregates;` in `lib.rs`. The `aggregates` module implementation can either be in `aggregates/mod.rs` (this file) or in `aggregates.rs` (which doesn't exist). This means that we can't rename `mod.rs` to `gapfill_delta.rs`. We could move the `GapfillDelta...` stuff into a separate source file named `gapfill_delta.rs`, but I'm not sure that that adds any value.

@bitwalker maybe you could add your thoughts here. I'm not a Rust guru.

**Comment by JamesGuthrie** [comment]
@cevian I looked into code coverage. It's possible to get a code coverage report. It does require using the `nightly` compiler though. I have added an issue for it.

File: src/aggregates/mod.rs
Comment (bitwalker) [reply src/aggregates/mod.rs:1]: As a general rule, I use the `<module>/mod.rs` convention whenever there are multiple submodules, and use `mod.rs` to hold module-global stuff (like custom error/result types, constants, etc), and to re-export submodule types/functions/etc., that the module aims to provide; it's also a good place to define any `macro_rules!` macros used within the module. For anything else, I use the single-file `<module>.rs` convention. 

The approach @JamesGuthrie used here feels right to me. I do agree though that the `GapfillDeltaTransition` bits probably make sense to define in a submodule; the rest of it seems like it is in the right place, at least for now.

File: src/aggregates/vector_selector.rs
Comment (JamesGuthrie) [reply src/aggregates/vector_selector.rs:77]: Based on the coverage report that I was able to generate, these functions are not being hit by the tests. I'm having some issues with debugging this at the moment though. Will keep digging.

File: src/aggregates/vector_selector.rs
Comment (bitwalker) [inline src/aggregates/vector_selector.rs:12]: We should probably change comments like these to be doc comments, i.e. `///` rather than `//`, and then control whether or not they show up in the generated crate docs using doc attributes, e.g. `#[doc(hidden)]`

**Review by bitwalker — APPROVED** [review]
I'm still getting caught up, but this LGTM!

File: src/aggregates/mod.rs
Comment (cevian) [reply src/aggregates/mod.rs:28]: ok let's just put in a code-comment todo to rename in the future

File: src/aggregates/vector_selector.rs
Comment (cevian) [reply src/aggregates/vector_selector.rs:77]: Have you tried breaking the logic and seeing if tests fail? It may be a misreport by the coverage tool

File: src/aggregates/mod.rs
Comment (cevian) [reply src/aggregates/mod.rs:1]: Ok I do think we should move GapfillDeltaTransition but will leave it up to you in this PR.

**Review by cevian — APPROVED** [review]
All this code is looking lots better.

File: src/aggregates/vector_selector.rs
Comment (JamesGuthrie) [reply src/aggregates/vector_selector.rs:77]: The coverage was correct. I managed to figure out how to force PG into performing a partial aggregation, and it surfaced a bug. Partial aggregation on `vector_select` is broken (also in previous releases - I tested `0.2.0` and `0.3.0` and has probably never worked. That might be interesting in itself, because obviously we're not generating queries in `promscale` that _would_ use it.

File: src/aggregates/vector_selector.rs
Comment (JamesGuthrie) [reply src/aggregates/vector_selector.rs:12]: Agree. I have done this already in some places. IMO it requires more than just converting a `//` to a `///`, as I think that the quality of the code documentation needs to be improved. I've opened #42 for this.

File: src/aggregates/vector_selector.rs
Comment (JamesGuthrie) [reply src/aggregates/vector_selector.rs:77]: And I found out how to fix it.

File: src/aggregates/mod.rs
Comment (JamesGuthrie) [reply src/aggregates/mod.rs]: done

**Review by cevian — APPROVED** [review]
LGTM

**Comment by bitwalker** [comment]
@JamesGuthrie I've rebased develop on latest master, and rebased your branch for this PR on top of develop. As mentioned in a different PR, I think we should try and avoid merging changes to master without going through develop first, and base all of our feature branches on develop (like you've done here), that way we can avoid needing to do the rebase dance like this in the future - but at least now we can merge your ongoing work into develop, and then merge develop into master when we're ready to cut a new release. If you have other local branches based on develop, just be aware that you'll need to rebase them on develop as well.

**Comment by JamesGuthrie** [comment]
Awesome, thanks for cleaning that up @bitwalker

**Pull request merged**
